File: algos/409-longestPalindrome/409-longestPalindrome.js
/**
 * @param {string} s
 * @return {number}
 */
var longestPalindrome = function(s) {
    // Build a Map of character counts
    const charMap = new Map();
    let result = 0;
    let oddFound = false;

    // Count the occurrences of each character
    for (const char of s) {
        charMap.set(char, (charMap.get(char) || 0) + 1);
    }

    // Calculate the length of the longest palindrome
    for (const [char, charCount] of charMap) {
        if (charCount % 2 === 0) {
            result += charCount;
        } else {
            oddFound = true;
            result += charCount - 1;
        }
    }

    // If an odd count is found, add one more character to the result
    if (oddFound) {
        result += 1;
    }

    return result;
};

File: algos/409-longestPalindrome/409-longestPalindrome.py
class Solution:
    def longestPalindrome(self, s: str) -> int:
        # Build a dictionary to store character counts
        char_count = {}
        
        # Count occurrences of each character
        for char in s:
            char_count[char] = char_count.get(char, 0) + 1
        
        # Initialize result and flag for odd count
        result = 0
        odd_found = False
        
        # Traverse the dictionary
        for count in char_count.values():
            # If count is even, add it directly to the result
            if count % 2 == 0:
                result += count
            # If count is odd, add count - 1 to result and set odd_found to True
            else:
                odd_found = True
                result += count - 1
        
        # If an odd count was found, add 1 to the result
        if odd_found:
            result += 1
        
        return result